File: chess/piece.py
from __future__ import annotations

class Square:
    '''Base class for all squares'''
    def __init__(self, colour: str):
        self.__type = ' '
        self.__colour = colour

    def is_colour(self, colour: str) -> bool:
        return self.__colour == colour

    def get_colour(self) -> str:
        if self.__colour:
            return self.__colour
        return ' '

    def get_type(self) -> str:
        return self.__type

    def colour_check(self, working: Square, target: Square, turn: str) -> bool: # NOTE - Can't type hint working or target here correctly for some reason
        if working.is_colour(turn):
            try:
                if not target.is_colour(turn):
                    return True
            except AttributeError:
                return True
        return False

class Pawn(Square):
    '''Class for the Pawn Piece'''
    def __init__(self, colour: str):
        super().__init__(colour)
        self.__moved = False

    def is_valid(self, working: Square, target: Square, working_index: int, target_index: int, turn: str, board: list) -> bool: # TODO en passant
        valid = False
        if Square.colour_check(self, working, target, turn):
            print("Correct colour chosen")
            if self.get_colour() == 'W': # Logic for white's movement
                if target_index == working_index + 16 and not self.__moved and target.getType() == ' ': # Logic for moving forwards 2 squares
                    valid = True
                elif target_index == working_index + 8 and target.getType() == ' ': # Logic for moving forwards 1 sqaure
                    valid = True
                elif target.get_colour() != self.get_colour() and target.get_colour() != ' ': # Logic for taking Square
                    if target_index == working_index + 7 or target_index == working_index + 9:
                        valid = True
            else: # Logic for black's movement
                if target_index == working_index - 16 and not self.__moved and target.getType() == ' ': # Logic for moving forward 2 squares
                    valid = True
                elif target_index == working_index - 8 and target.getType() == ' ': # Logic for moving forward one square
                    valid = True
                elif target.get_colour() != self.get_colour() and target.get_colour() != ' ': # Logic for taking a Square
                    if target_index == working_index - 7 or target_index == working_index - 9:
                        valid = True
        # Check for en passant
        
        if valid and not self.__moved:
            self.__moved = True
        return valid

    def get_type(self) -> str:
        return 'P'

class Knight(Square):
    '''Class for the Knight Piece'''
    def __init__(self, colour):
        super().__init__(colour)

    def is_valid(self, working: Square, target: Square, working_index: int, target_index: int, turn: str, board: list) -> bool:
        if Square.colour_check(self, working, target, turn):
            if abs(working_index - target_index) in [17, 15, 6, 10] and \
                target_index != working_index:
                return True
        return False
    
    def get_type(self):
        return 'N'

class Bishop(Square):
    '''Class for the Bishop Piece'''
    def __init__(self, colour):
        super().__init__(colour)

    def is_valid(self, working: Square, target: Square, working_index: int, target_index: int, turn: str, board: list) -> bool:
        '''Checks if the move is valid for bishops'''
        if Square.colour_check(self, working, target, turn):
            direction = working_index - target_index
            if abs(direction) % 7 == 0:
                for i in range(7, abs(direction), 7):
                    if board[working_index + i if direction > 0 else working_index - i]:
                        return False
                return True
            elif abs(direction) % 9 == 0:
                for i in range(9, abs(direction), 9):
                    if board[working_index + i if direction > 0 else working_index - i]:
                        return False
                return True
        return False

    def get_type(self):
        return 'B'

class Rook(Square):
    '''Class for the Rook Piece'''
    def __init__(self, colour):
        super().__init__(colour)
        self.has_moved = False

    def is_valid(self, working: Square, target: Square, working_index: int, target_index: int, turn: str, board: list) -> bool:
        if Square.colour_check(self, working, target, turn):
            direction = working_index - target_index
            if abs(direction) % 8 == 0:
                for i in range(8, abs(direction), 8):
                    if board[working_index + i if direction > 0 else working_index - i]:
                        return False
            elif abs(direction) % 1 == 0:
                for i in range(1, abs(direction), 1):
                    if board[working_index + i if direction > 0 else working_index - i]:
                        return False
            return True
        return False
        # Check for piece collisions in the move
        
    def get_type(self):
        return 'R'
        

class Queen(Square):
    '''Class for the Queen Piece'''
    def __init__(self, colour):
        super().__init__(colour)

    def is_valid(self, working: Square, target: Square, working_index: int, target_index: int, turn: str, board: list) -> bool:
        if Square.colour_check(self, working, target, turn):
            direction = working_index - target_index
            if abs(direction) % 7 == 0:
                for i in range(7, abs(direction), 7):
                    if board[working_index + i if direction > 0 else working_index - i]:
                        return False
                return True
            elif abs(direction) % 9 == 0:
                for i in range(9, abs(direction), 9):
                    if board[working_index + i if direction > 0 else working_index - i]:
                        return False
                return True
            elif abs(direction) % 8 == 0:
                for i in range(8, abs(direction), 8):
                    if board[working_index + i if direction > 0 else working_index - i]:
                        return False
                return True
        return False

    def get_type(self):
        return 'Q'

class King(Square):
    '''Class for the King Piece'''
    def __init__(self, colour):
        super().__init__(colour)
        self.has_moved = False

    def is_valid(self, working: Square, target: Square, working_index: int, target_index: int, turn: str) -> bool:
        if Square.colour_check(self, working, target, turn):
            if abs(working_index - target_index) in [15, 1, 17, 9, 7, 8, 16, 24]:
                return False
    
    def get_type(self):
        return 'K'

class Empty(Square):
    '''Class for Empty spaces on the board'''
    def __init__(self):
        Square.__init__(self, ' ')

    def get_type(self):
        return ' '

    def get_colour(self):
        return ' '

if __name__ == '__main__':
    print("File run incorrectly - please run game.py instead")

File: chess-rust/src/asset/piece.rs
pub enum Piece {
    Pawn(Data, bool),
    Rook(Data, bool),
    Knight(Data),
    Bishop(Data),
    Queen(Data),
    King(Data, bool),
    Empty(Data),
}

impl Piece {
    fn is_valid(&self, start: usize, end: usize, turn: char, board: &Vec<Piece>) -> bool {
        self.is_valid(start, end, turn, board)
    }

    fn get_type(&self) -> char {
        self.get_type()
    }

    fn get_colour(&self) -> char {
        self.get_colour()
    }

    pub fn new_piece(piece: char, colour: char) -> Piece {
        match piece {
            'P' => Pawn::new(colour),
            'R' => Rook::new(colour),
            'N' => Knight::new(colour),
            'B' => Bishop::new(colour),
            'Q' => Queen::new(colour),
            'K' => King::new(colour),
            ' ' => Empty::new(),
            _ => panic!("Invalid piece"),
        }
    }
}

#[derive(Clone)]
struct Data {
    piece: char,
    colour: char,
}

struct Pawn {
    data: Data,
    moved: bool,
}

impl Pawn {
    fn new(colour: char) -> Piece {
        Piece::Pawn(Data { piece: 'P', colour }, false)
    }

    fn is_valid(
        &self,
        working_index: usize,
        target_index: usize,
        turn: char,
        _board: &[Piece],
    ) -> bool {
        let working_row = working_index / 8;
        let working_col = working_index % 8;
        let target_row = target_index / 8;
        let target_col = target_index % 8;

        let mut valid = false;

        if self.is_colour(turn) {
            if self.is_colour('W') {
                if working_row == 1 {
                    if target_row == 3 && target_col == working_col {
                        valid = true;
                    }
                } else if target_row == working_row + 1 && target_col == working_col {
                    valid = true;
                }
            } else {
                if working_row == 6 {
                    if target_row == 4 && target_col == working_col {
                        valid = true;
                    }
                } else if target_row == working_row - 1 && target_col == working_col {
                    valid = true;
                }
            }
        }

        valid
    }

    fn is_colour(&self, colour: char) -> bool {
        self.get_colour() == colour
    }

    fn get_colour(&self) -> char {
        self.data.colour
    }
}

struct Rook {
    data: Data,
    moved: bool,
}

impl Rook {
    fn new(colour: char) -> Piece {
        Piece::Rook(Data { piece: 'R', colour }, false)
    }

    fn is_valid(
        &self,
        working_index: usize,
        target_index: usize,
        turn: char,
        board: &[Piece],
    ) -> bool {
        let working_row = working_index / 8;
        let working_col = working_index % 8;
        let target_row = target_index / 8;
        let target_col = target_index % 8;

        let mut valid = false;

        if self.is_colour(turn) {
            if working_row == target_row {
                valid = true;
            } else if working_col == target_col {
                valid = true;
            }
        }

        // Check if piece is in way
        if valid {
            if working_row == target_row {
                let mut min = working_col;
                let mut max = target_col;

                if min > max {
                    min = target_col;
                    max = working_col;
                }

                for i in min..max {
                    if board[i + working_row * 8].get_type() != ' ' {
                        valid = false;
                    }
                }
            } else if working_col == target_col {
                let mut min = working_row;
                let mut max = target_row;

                if min > max {
                    min = target_row;
                    max = working_row;
                }

                for i in min..max {
                    if board[working_col + i * 8].get_type() != ' ' {
                        valid = false;
                    }
                }
            }
        }

        valid
    }

    fn is_colour(&self, colour: char) -> bool {
        self.get_colour() == colour
    }

    fn get_colour(&self) -> char {
        self.data.colour
    }
}

struct Knight {
    data: Data,
}

impl Knight {
    fn new(colour: char) -> Piece {
        Piece::Knight(Data { piece: 'N', colour })
    }

    fn is_valid(
        &self,
        working_index: usize,
        target_index: usize,
        turn: char,
        _board: &[Piece],
    ) -> bool {
        let working_row = working_index / 8;
        let working_col = working_index % 8;
        let target_row = target_index / 8;
        let target_col = target_index % 8;

        let mut valid = false;

        if self.is_colour(turn) {
            if (working_row - target_row) == 2 && (working_col - target_col) == 1 {
                valid = true;
            } else if (working_row - target_row) == 1 && (working_col - target_col) == 2 {
                valid = true;
            }
        }

        valid
    }

    fn is_colour(&self, colour: char) -> bool {
        self.get_colour() == colour
    }

    fn get_colour(&self) -> char {
        self.data.colour
    }
}

struct Bishop {
    data: Data,
}

impl Bishop {
    fn new(colour: char) -> Piece {
        Piece::Bishop(Data { piece: 'B', colour })
    }

    fn is_valid(
        &self,
        working_index: usize,
        target_index: usize,
        turn: char,
        board: &[Piece],
    ) -> bool {
        let working_row = working_index / 8;
        let working_col = working_index % 8;
        let target_row = target_index / 8;
        let target_col = target_index % 8;

        let mut valid = false;

        if self.is_colour(turn) {
            if (working_row - target_row) == (working_col - target_col) {
                valid = true;
            }
        }

        // Check if piece is in way
        if valid {
            let mut min_row = working_row;
            let mut max_row = target_row;
            let mut min_col = working_col;
            let max_col = target_col;

            if min_row > max_row {
                min_row = target_row;
                max_row = working_row;
            }

            if min_col > max_col {
                min_col = target_col;
            }

            for i in 1..(max_row - min_row) {
                if board[(min_col + i) + (min_row + i) * 8].get_type() != ' ' {
                    valid = false;
                }
            }
        }

        valid
    }

    fn is_colour(&self, colour: char) -> bool {
        self.get_colour() == colour
    }

    fn get_colour(&self) -> char {
        self.data.colour
    }
}

struct Queen {
    data: Data,
}

impl Queen {
    fn new(colour: char) -> Piece {
        Piece::Queen(Data { piece: 'Q', colour })
    }

    fn is_valid(
        &self,
        working_index: usize,
        target_index: usize,
        turn: char,
        board: &[Piece],
    ) -> bool {
        let working_row = working_index / 8;
        let working_col = working_index % 8;
        let target_row = target_index / 8;
        let target_col = target_index % 8;

        let mut valid = false;

        if self.is_colour(turn) {
            if working_row == target_row {
                valid = true;
            } else if working_col == target_col {
                valid = true;
            } else if (working_row - target_row) == (working_col - target_col) {
                valid = true;
            }
        }

        // Check if piece is in way
        if valid {
            if working_row == target_row {
                let mut min = working_col;
                let mut max = target_col;

                if min > max {
                    min = target_col;
                    max = working_col;
                }

                for i in min..max {
                    if board[i + working_row * 8].get_type() != ' ' {
                        valid = false;
                    }
                }
            } else if working_col == target_col {
                let mut min = working_row;
                let mut max = target_row;

                if min > max {
                    min = target_row;
                    max = working_row;
                }

                for i in min..max {
                    if board[working_col + i * 8].get_type() != ' ' {
                        valid = false;
                    }
                }
            } else if (working_row - target_row) == (working_col - target_col) {
                let mut min_row = working_row;
                let mut max_row = target_row;
                let mut min_col = working_col;
                let max_col = target_col;

                if min_row > max_row {
                    min_row = target_row;
                    max_row = working_row;
                }

                if min_col > max_col {
                    min_col = target_col;
                }

                for i in 1..(max_row - min_row) {
                    if board[(min_col + i) + (min_row + i) * 8].get_type() != ' ' {
                        valid = false;
                    }
                }
            }
        }

        valid
    }

    fn is_colour(&self, colour: char) -> bool {
        self.get_colour() == colour
    }

    fn get_colour(&self) -> char {
        self.data.colour
    }
}

struct King {
    data: Data,
    moved: bool,
}

impl King {
    fn new(colour: char) -> Piece {
        Piece::King(Data { piece: 'K', colour }, false)
    }

    fn is_valid(
        &self,
        working_index: usize,
        target_index: usize,
        turn: char,
        _board: &[Piece],
    ) -> bool {
        let working_row = working_index / 8;
        let working_col = working_index % 8;
        let target_row = target_index / 8;
        let target_col = target_index % 8;

        let mut valid = false;

        if self.is_colour(turn) {
            if (working_row - target_row) <= 1 && (working_col - target_col) <= 1 {
                valid = true;
            }
        }

        valid
    }

    fn is_colour(&self, colour: char) -> bool {
        self.get_colour() == colour
    }

    fn get_colour(&self) -> char {
        self.data.colour
    }
}

struct Empty {
    data: Data,
}

impl Empty {
    fn new() -> Piece {
        Piece::Empty(Data {
            piece: ' ',
            colour: ' ',
        })
    }
}
